### PR TITLE
CORDA-3071 - Checkpoint agent tool

### PR DESCRIPTION
**Pull request by josecoll**
The Checkpoint Agent is a new tool that provides checkpoint stack sizing diagnostics information.
Based on the experimental Kryo hook, the checkpoint agent intercepts Kryo flow checkpoint reads and writes, and prints a hierarchical view of the entire Flow Object stack with sizing information at each level.
May be used standalone and/or in conjunction with the recent CRaSH ``dumpCheckpoints`` command. https://github.com/corda/corda/pull/5171

This tool was built following a troubleshooting event with a real corda network.

### CODE REVIEW

File: tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+        fun parseArguments(argumentsString: String?) {
+            argumentsString?.let {
+                val nvpList = it.split(",")
+                nvpList.forEach {
+                    val nvpItem = it.split("=")
+                    if (nvpItem.size == 2) {
+                        when (nvpItem[0].trim()) {
+                            "instrumentClassname" -> instrumentClassname = nvpItem[1]
+                            "instrumentType" -> try { instrumentType = InstrumentationType.valueOf(nvpItem[1].toUpperCase()) } catch (e: Exception) {
+                                println("Invalid value: ${nvpItem[1]}. Please specify read or write.")
```
Comment (rick-r3) [inline tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt]: Do all these `println` need to prefix themselves with `CheckpointAgent:` or something so it's more obvious where the message originates for someone seeing it ?

Maybe user a helper method to standardise it.  Should it go to `stderr` instead?

File: tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+    }
+
+    private fun instrumentClass(clazz: CtClass): CtClass? {
+        for (method in clazz.declaredBehaviors) {
+            if (instrumentType == CheckpointAgent.InstrumentationType.WRITE) {
+                if (method.name == "write") {
+                    val parameterTypeNames = method.parameterTypes.map { it.name }
+                    if (parameterTypeNames == listOf("com.esotericsoftware.kryo.Kryo", "com.esotericsoftware.kryo.io.Output", "java.lang.Object")) {
+                        if (method.isEmpty) continue
+                        log.debug("Instrumenting on write: ${clazz.name}")
```
Comment (rick-r3) [inline tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt]: do you have access to our lazy debug extension `log.debug { "message" }` ?  If not, just have a local one in this file.  Applies throughout this class.  Building these strings is expensive and should be protected by a `isDebugEnabled` condition as our lazy implementation does.

File: tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+                val arrayValue = value as BooleanArray
+                log.debug("readFieldExit boolean array: $clazz = ${arrayValue.joinToString(",")}")
+                return arrayValue.joinToString(",")
+            }
+            log.debug("ARRAY OF TYPE: $clazz (size: ${(value as Array<Any?>).size})")
+        }
+        return null
+    }
+
+    // https://stackoverflow.com/questions/9655181/how-to-convert-a-byte-array-to-a-hex-string-in-java
```
Comment (rick-r3) [inline tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt]: Remove this comment.  There's copyright issues with copying from stack overflow.  But we could claim this was obvious without the comment.

File: tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+        val DEFAULT_MINIMUM_SIZE = 8 * 1024
+        val DEFAULT_MAXIMUM_SIZE = 1000 * 1024
+        val DEFAULT_STACK_DEPTH = 12
+
+        // startup arguments
+        var instrumentClassname = DEFAULT_INSTRUMENT_CLASSNAME
+        var minimumSize = DEFAULT_MINIMUM_SIZE
+        var maximumSize = DEFAULT_MAXIMUM_SIZE
+        var instrumentType = DEFAULT_INSTRUMENT_TYPE
+        var stackDepth = DEFAULT_STACK_DEPTH
```
Comment (rick-r3) [inline tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt]: `graphDepth`

File: tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+        WRITE, READ
+    }
+
+    companion object {
+        // custom argument defaults
+        val DEFAULT_INSTRUMENT_CLASSNAME = "net.corda.node.services.statemachine.FlowStateMachineImpl"
+        val DEFAULT_INSTRUMENT_TYPE = InstrumentationType.READ
+        val DEFAULT_MINIMUM_SIZE = 8 * 1024
+        val DEFAULT_MAXIMUM_SIZE = 1000 * 1024
+        val DEFAULT_STACK_DEPTH = 12
```
Comment (rick-r3) [inline tools/checkpoint-agent/src/main/kotlin/net/corda/tools/CheckpointAgent.kt]: `graphDepth`

File: docs/source/tools-checkpoint-agent.rst [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+
+.. code-block:: shell
+
+    checkpoint-agent.jar=[instrumentType=<read|write>],[instrumentClassname=<CLASSNAME>],[minimumSize=<MIN_SIZE>],[maximumSize=<MAX_SIZE>, [stackDepth=<DEPTH>], [printOnce=<true|false>]
+
+* ``instrumentType``: whether to output checkpoints on read or write. Possible values: [read, write]. Default: read.
+* ``instrumentClassname``: specify the base type of objects to log. The default setting is to process all *Flow* object types. Default: net.corda.node.services.statemachine.FlowStateMachineImpl.
+* ``minimumSize``: specifies the minimum size (in bytes) of objects to log. Default: 8192 bytes
+* ``maximumSize``: specifies the maximum size (in bytes) of objects to log. Default: 1024000 bytes
+* ``stackDepth``: specifies how many levels deep to display the stack output. Default: 12
```
[the remark this answers is not in the document]
Comment (josecoll) [reply docs/source/tools-checkpoint-agent.rst]: yes - it is the nesting depth of objects in the graph. Will change.

File: node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt [hunk 7]
```diff
@@ -116,6 +126,28 @@ class CheckpointDumper(private val checkpointStorage: CheckpointStorage, private
...
+        }
+        catch (e: Exception) {
+            log.error("Checkpoint agent instrumentation failed for checkpointId: $checkpointId\n. ${e.message}")
+        }
+    }
+
+    private fun checkpointAgentRunning(): Boolean {
+        val agentProperties = VMSupport.getAgentProperties()
+        return agentProperties.values.any { value ->
+            (value is String && value.contains("checkpoint-agent.jar"))
```
[the remark this answers is not in the document]
Comment (josecoll) [reply node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: Added warning to the effect:

warning:: The checkpoint agent JAR file must be called "checkpoint-agent.jar" as the checkpoint dump support code uses Java reflection to
   determine whether the VM has been instrumented or not at runtime.

**Pull request merged**